[ConstraintSystem] Strip labels from enum elements when passed as values

xedin opened the pull request:
Enum elements have to be treated the same way as regular functions
when passed as values, which means labels have to be stripped from
their argument types.

Resolves: rdar://problem/32300339.

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[ConstraintSystem] Strip labels from enum elements when passed as values

Enum elements have to be treated the same way as regular functions
when passed as values, which means labels have to be stripped from
their argument types.

Resolves: rdar://problem/32300339.

xedin commented:
/cc @slavapestov please take a look and let me know what you think

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please test source compatibility

rudkx commented on lib/Sema/ConstraintSystem.cpp:
```diff
@@ -712,30 +712,54 @@ void ConstraintSystem::recordOpenedTypes(
+  //
+  // `E.foo` has to act as a regular function type passed as a value.
+  if (!TC.getLangOpts().isSwiftVersion3()) {
+    if (auto *EED = dyn_cast<EnumElementDecl>(decl)) {
```
It's safer to have this under the version check, but is it actually necessary?

What I'm wondering is if there might be cases where the Swift 3 behavior is also broken.

slavapestov replied:
The Swift 3 behavior is broken if your enum argument type has multiple elements.

rudkx reviewed (approved):
Otherwise LGTM.

slavapestov commented on lib/Sema/ConstraintSystem.cpp:
```diff
@@ -712,30 +712,54 @@ void ConstraintSystem::recordOpenedTypes(
                                             FunctionRefKind functionRefKind) {
+  unsigned numParameterLists = 0;
+
+  // This is required because if enum case has a function type and passed
```
Please either remove this comment, or rewrite it to talk about the common behavior of both functions and enum cases.

slavapestov commented on test/Constraints/diagnostics.swift:
```diff
@@ -977,4 +977,14 @@ let _: KeyPath<R32101765, Float> = \R32101765.prop32101765
\ No newline at end of file
+// expected-error@-1 {{type 'Int' has no member 'unknown'}}
+
+// rdar://problem/32300339 - enum elements with labeled argument should be treated as functions when passed as value
```
Please put this in a new file, test/Compatibility/enum_cases.swift.

slavapestov commented on test/Constraints/diagnostics_swift4.swift:
```diff
@@ -44,4 +44,14 @@ class R<K: Hashable, V> {
\ No newline at end of file
+}
+
+// rdar://problem/32300339 - enum elements with labeled argument should be treated as functions when passed as value
```
Please put this in a new file, test/Constraints/enum_cases.swift.

Both files need a lot more tests, for example it would be nice to test:

- case has no label
- case has label
- case is a tuple with multiple elements, some of which have labels
- assign case to a 'let' binding with an explicit type

slavapestov commented:
I wonder if we can get rid of scalar-to-tuple conversions entirely in Swift 4 mode once this goes in.

xedin commented:
@slavapestov @rudkx I've addressed all of the notes - reformatted the comment and added a bunch more tests to the {Compatibility, Constraints}/enum_cases.swift . Also added a test which show the different in behavior between swift 3 and 4.

slavapestov commented:
LGTM!

rudkx commented:
Looks great!

xedin commented:
@swift-ci please test and merge

slavapestov commented:
@xedin We should get this into beta 2.

xedin commented:
@slavapestov Will do!
